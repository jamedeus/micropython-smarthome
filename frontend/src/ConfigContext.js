import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';
import { get_config_template } from 'util/metadata';
import { sleep } from 'util/helper_functions';
import { v4 as uuid } from 'uuid';
import { get_instance_metadata, ir_keys } from 'util/metadata';
import { api_target_options, target_node_ip } from 'util/django_util';


// Takes object and key prefix, returns object with all keys that begin with prefix
function filterObject(obj, prefix) {
    return Object.entries(obj).reduce((acc, [key, value]) => {
        if (key.startsWith(prefix)) {
            acc[key] = value;
        }
        return acc;
    }, {});
}

// Takes object and key prefix, returns array of keys that begin with prefix
function filterObjectKeys(obj, prefix) {
    return Object.keys(filterObject(obj, prefix));
}

export const ConfigContext = createContext();

export const ConfigProvider = ({ children }) => {
    // Load config context set by django template
    const [config, setConfig] = useState(() => {
        // Context element created with json_script django filter
        const element = document.getElementById("config");

        let config;
        // Use context if present, otherwise use blank template
        if (element) {
            config = JSON.parse(element.textContent);
        } else {
            config = {
                metadata: {
                    id: '',
                    floor: '',
                    location: '',
                    schedule_keywords: {}
                },
                wifi: {
                    ssid: '',
                    password: ''
                }
            };
        }

        // Check for api-target with invalid target IP (doesn't match any node)
        // If found, clear ip, default_rule, and schedule rules fields
        const apiTargets = Object.entries(config).filter(([id, params]) => {
            return id.startsWith('device') && params._type === 'api-target';
        });
        apiTargets.forEach(([id, params]) => {
            if (!Object.values(api_target_options.addresses).includes(params.ip)) {
                config[id]['ip'] = '';
                config[id]['default_rule'] = '';
                config[id]['schedule'] = {};
            }
        });

        return config;
    });

    // Create state to control card delete animations
    // Card matching ID fades out, cards in category with higher index slide up
    const [deleteing, setDeleteing] = useState({id: '', category: '', index: ''});

    const updateConfig = newConfig => {
        setConfig(newConfig);
    };

    // Create state for device and sensor UUIDs (used as react key)
    // Pre-populate with id-UUID pairs for existing instances (if editing)
    const [UUIDs, setUUIDs] = useState(
        {
            device: Object.fromEntries(
                filterObjectKeys(config, "device").map((id) => [id, uuid()])
            ),
            sensor: Object.fromEntries(
                filterObjectKeys(config, "sensor").map((id) => [id, uuid()])
            )
        }
    );

    // Takes card index, returns UUID
    const getKey = (id, category) => {
        return UUIDs[category][id];
    };

    // Create state object to control invalid field highlight
    const [highlightInvalid, setHighlightInvalid] = useState(false);

    // Handler for add device and add sensor buttons
    const addInstance = (category) => {
        // Get index of new instance
        const index = Object.keys(filterObject(config, category)).length + 1;

        // Add key to state object with empty config
        // Will be populated with metadata template when user selects type
        // _type = undefined used to detect empty and add invalid highlight
        setConfig({ ...config, [`${category}${index}`]: {'_type': undefined} });

        // Create UUID for new instance
        setUUIDs({ ...UUIDs, [category]: {
            ...UUIDs[category], [`${category}${index}`]: uuid()
        }});
    };

    // Handler for delete button on device and sensor cards
    const deleteInstance = async (id) => {
        // Get height of deleted card + 1.5rem (margin between cards)
        const card = document.getElementById(`${id}-card`);
        const remPx = parseFloat(getComputedStyle(document.documentElement).fontSize);
        const animation_height = card.clientHeight / remPx + 1.5;

        // Set CSS var used in delete animation
        document.documentElement.style.setProperty(
            '--animation-height',
            `${animation_height}rem`
        );

        // Start animation, card being deleted will fade out, all cards below
        // will slide up to fill empty space (uses --animation-height)
        setDeleteing({
            id: id,
            category: id.replace(/[0-9]/g, ''),
            index: parseInt(id.replace(/[a-zA-z]/g, ''))
        });

        // Wait for animation to complete
        await sleep(800);

        // Remove instance from config state, reset animation state
        const newConfig = { ...config };
        delete newConfig[id];
        setConfig(update_ids(id, newConfig));
        setDeleteing({id: '', category: '', index: ''});
    };

    // Handler for type select dropdown in device and sensor cards
    const changeInstanceType = (id, category, event) => {
        const template = get_config_template(category, event.target.value);
        setConfig({ ...config, [id]: template});
    };

    // Handler for all inputs inside device and sensor cards
    // Takes device/sensor ID, param, and value; updates state object
    const handleInputChange = (id, param, value) => {
        const update = { ...config[id], [param]: value };
        setConfig({ ...config, [id]: update});
    };

    // Updates multiple instance params in a single event
    // Takes device/sensor ID, full dict of params; overwrites state object section
    const handleInstanceUpdate = (id, params) => {
        setConfig({ ...config, [id]: params});
    };

    // Handler for sensor target select checkboxes
    const handleSensorTargetSelect = (sensor, target, checked) => {
        // Copy config section
        const update = { ... config[sensor] };
        // Add target if not already present
        if (checked) {
            if (update["targets"].indexOf(target) === -1) {
                update["targets"].push(target);
            }
        // Remove existing target if present
        } else {
            update["targets"] = update["targets"].filter(existing => existing !== target);
        }
        setConfig({ ...config, [sensor]: update });
    };

    // Handler for IR target checkboxes
    const handleIrTargetSelect = (target, checked) => {
        const ir_blaster = { ...config.ir_blaster };
        // Add target if not already present
        if (checked) {
            if (ir_blaster.target.indexOf(target) === -1) {
                ir_blaster.target.push(target);
            }
        // Remove existing target if present
        } else {
            ir_blaster.target = ir_blaster.target.filter(existing => existing !== target);
        }
        setConfig({ ...config, ['ir_blaster']: ir_blaster });
    };

    // Called by deleteInstance, decrements IDs of all subsequent instances to prevent gaps
    // Example: If device2 is deleted, device3 becomes device2, device4 becomes device3, etc
    function update_ids(target, state) {
        // Get category (device or sensor) and index of removed instance
        const category = target.replace(/[0-9]/g, '');
        const index = target.replace(/[a-zA-Z]/g, '');

        // Get list of all instances in same category
        const instances = filterObject(state, category);

        // Get UUIDs of all instances in same category
        const newUUIDs = { ...UUIDs[category] };

        // Get list of all sensors (used to update target IDs)
        const sensors = filterObject(state, 'sensor');

        // If target is device remove from all sensor target lists
        if (category === 'device') {
            for (const sensor in sensors) {
                // Prevent error if sensor type not selected yet (no targets key)
                if (state[sensor]['targets']) {
                    state[sensor]['targets'] = state[sensor]['targets'].filter(
                        item => item !== target
                    );
                }
            }
        }

        // Iterate all instances in category starting from the removed instance index
        for (let i=parseInt(index); i<Object.entries(instances).length+1; i++) {
            // Removed index now available, decrement next index by 1
            const new_id = `${category}${i}`;
            const old_id = `${category}${i+1}`;
            state[new_id] = JSON.parse(JSON.stringify(state[old_id]));
            delete state[old_id];

            // Decrement UUID index to keep associated with correct card
            newUUIDs[new_id] = newUUIDs[old_id];
            delete newUUIDs[old_id];

            // Decrement device index in sensor targets lists to match above
            if (category === 'device') {
                for (const sensor in sensors) {
                    if (state[sensor]['targets'].includes(old_id)) {
                        state[sensor]['targets'] = state[sensor]['targets'].filter(
                            item => item !== old_id
                        );
                        state[sensor]['targets'].push(new_id);
                    }
                }
            }
        }

        // Update UUIDs
        setUUIDs({ ...UUIDs, [category]: {
            ...newUUIDs
        }});

        // Return state (calling function updates)
        return state;
    }

    // Takes ApiTarget target node IP, returns object containing all valid
    // options for target (used to populate ApiTargetRuleModal dropdowns)
    const getTargetNodeOptions = (ip) => {
        // Generate options from current state if self-targeting
        if (ip === '127.0.0.1' || ip === target_node_ip) {
            return getSelfTargetOptions();
        }

        // Find target node options in object if not self-targeting
        const friendly_name = Object.keys(api_target_options.addresses).find(key =>
            api_target_options.addresses[key] === ip
        );
        // Throw error if not found (prevent crash when opening modal)
        if (!friendly_name) {
            throw new Error(
                'getTargetNodeOptions received an IP that does not match an existing node'
            );
        }

        return api_target_options[friendly_name];
    };

    // Returns ApiTargetRuleModal options for all devices and sensors in config
    const getSelfTargetOptions = () => {
        const options = {};

        // Supported by all devices and sensors
        const universalOptions = [
            'enable',
            'disable',
            'enable_in',
            'disable_in',
            'set_rule',
            'reset_rule'
        ];

        // Add options for each configured device and sensor
        const devices = filterObject(config, "device");
        Object.entries(devices).forEach(([device, params]) => {
            // Add display string and universal options
            options[device] = {
                display: `${params.nickname} (${params._type})`,
                options: [ ...universalOptions ]
            };

            // Add on/off for all types except ApiTarget (prevent infinite loop)
            if (params._type !== 'api-target') {
                options[device]['options'].push('turn_on', 'turn_off');
            }
        });
        const sensors = filterObject(config, "sensor");
        Object.entries(sensors).forEach(([sensor, params]) => {
            // Add display string and universal options
            options[sensor] = {
                display: `${params.nickname} (${params._type})`,
                options: [ ...universalOptions ]
            };

            // Add trigger_sensor endpoint for triggerable sensors
            const metadata = get_instance_metadata('sensor', params._type);
            if (metadata.triggerable) {
                options[sensor]['options'].push('trigger_sensor');
            }
        });

        // If IR Blaster with at least 1 target configured, add options for each target
        if (Object.keys(config).includes("ir_blaster") && config.ir_blaster.target.length) {
            options['ir_key'] = {
                display: "Ir Blaster",
                options: [ ...config.ir_blaster.target ],
                keys: {}
            };
            config.ir_blaster.target.forEach(target => {
                options.ir_key.keys[target] = ir_keys[target];
            });
        }

        return options;
    };

    return (
        <ConfigContext.Provider value=
            {{
                config,
                updateConfig,
                getKey,
                highlightInvalid,
                setHighlightInvalid,
                addInstance,
                deleteInstance,
                changeInstanceType,
                handleInputChange,
                handleInstanceUpdate,
                handleSensorTargetSelect,
                handleIrTargetSelect,
                getTargetNodeOptions,
                deleteing
            }}
        >
            {children}
        </ConfigContext.Provider>
    );
};

ConfigProvider.propTypes = {
    children: PropTypes.node,
};

export { filterObject };
