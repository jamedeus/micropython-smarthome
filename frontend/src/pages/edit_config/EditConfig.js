import React, { useState, useContext } from 'react';
import { ConfigContext } from 'root/ConfigContext';
import Button from 'react-bootstrap/Button';
import { send_post_request, edit_existing, target_node_ip } from 'util/django_util';
import Page1 from './Page1';
import Page2 from './Page2';
import Page3 from './Page3';
import ApiTargetRuleModal from 'modals/ApiTargetRuleModal';
import { ErrorModalContext, ErrorModal } from 'modals/ErrorModal';
import UploadModal, { uploadConfigFile } from 'modals/UploadModal';

// Redirect back to overview page
function returnToOverview() {
    window.location.replace("/config_overview");
}

// Takes current config (state) object, compares with original from django
// context, returns True if changes have been made, false if no changes
function configModified(config) {
    // Get original config, compare with
    const original_config = JSON.parse(
        document.getElementById("config").textContent
    );
    return !areObjectsEqual(config, original_config);
}

// Takes 2 objects, recursively compares all subkeys, returns True if identical
// Used to detect unsaved changes, show warning before returning to overview
function areObjectsEqual(obj1, obj2) {
    // Direct comparison for sub-keys on recursive calls
    if (obj1 === obj2) {
        return true;
    }

    // Not equal if failed comparison above and either is not object
    if (typeof(obj1) !== 'object' || typeof(obj2) !== 'object') {
        return false;
    }

    const keys1 = Object.keys(obj1);
    const keys2 = Object.keys(obj2);

    // Not equal if different number of keys
    if (keys1.length !== keys2.length) {
        return false;
    }

    // Confirm contain same keys
    // Call recursively with values of each key in both objects
    for (let key of keys1) {
        if (!keys2.includes(key) || !areObjectsEqual(obj1[key], obj2[key])) {
            return false;
        }
    }

    return true;
}

// Takes config (state) object, returns true if any keys are empty strings
// Used to detect empty fields before changing page
function hasEmptyFields(config) {
    for (let key in config) {
        if (typeof(config[key]) === 'object') {
            if (hasEmptyFields(config[key])) {
                console.log(`Empty field in ${key}`);
                return true;
            }
        } else if ([null, '', undefined].includes(config[key])) {
            console.log(`Empty field: ${key}`);
            return true;
        }
    }
    return false;
}

// Takes config object, returns true if any keys are null or empty strings
// Used to detect unset schedule rule time before submitting
function hasEmptyKeys(config) {
    for (let key in config) {
        if ([null, '', undefined].includes(key)) {
            console.log('Config contains empty key');
            return true;
        } else if (typeof(config[key]) === 'object') {
            if (hasEmptyKeys(config[key])) {
                console.log('Config contains empty key');
                return true;
            }
        }
    }
}

const EditConfig = () => {
    // Set default page, get callback to change visible page
    const [page, setPage] = useState(1);

    // Get full config (state object)
    const { config, setHighlightInvalid } = useContext(ConfigContext);

    // Get state and callback for error modal
    const {
        errorModalContent,
        setErrorModalContent
    } = useContext(ErrorModalContext);

    function prevPage() {
        // Go back to overview if current page is page 1
        if (page === 1) {
            // Show unsaved changes warning if user modified any inputs
            if (configModified(config)) {
                setErrorModalContent({
                    ...errorModalContent,
                    ["visible"]: true,
                    ["title"]: "Warning",
                    ["error"]: "unsaved_changes",
                    ["handleConfirm"]: returnToOverview
                });
            // Go directly to overview if no unsaved changes
            } else {
                returnToOverview();
            }
        // Otherwise go to previous page
        } else {
            setPage(page - 1);
        }
    }

    function nextPage() {
        // Don't go to page2 if empty inputs exist on page1
        if (page === 1 && hasEmptyFields(config)) {
            setHighlightInvalid(true);
            return;
        }
        // Clear highlight, go to next page
        setHighlightInvalid(false);
        setPage(page + 1);
    }

    function friendlyNameToFilename(friendlyName) {
        return `${friendlyName.toLowerCase().replaceAll(' ', '-')}.json`;
    }

    // Post full config (state object) to backend when submit clicked
    async function submitButton() {
        console.log(config);

        // Don't submit if config has empty keys (schedule rule time not set)
        if (hasEmptyKeys(config)) {
            setHighlightInvalid(true);
            return;
        }

        // Overwrites if editing existing config, otherwise create config
        const response = await send_post_request(
            edit_existing ? "generate_config_file/True" : "generate_config_file",
            config
        );

        // If successfully created new config, redirect to overview
        if (!edit_existing && response.ok) {
            // Redirect back to overview where user can upload the new config
            returnToOverview();

        // If successfully edited existing config, re-upload to target node
        } else if (edit_existing && response.ok) {
            // Convert friendly name into config filename
            const target_filename = friendlyNameToFilename(config.metadata.id);
            // Show upload modal, upload, redirect to overview when complete
            uploadConfigFile(target_filename, target_node_ip, true, returnToOverview);

        // If config with same name already exists, show overwrite prompt
        } else if (!edit_existing && response.status == 409) {
            setErrorModalContent({
                ...errorModalContent,
                ["visible"]: true,
                ["title"]: "Duplicate Warning",
                ["error"]: "duplicate",
                ["body"]: config.metadata.id,
                ["handleConfirm"]: confirmOverwriteDuplicate
            });

        // If other error, display in alert
        } else {
            alert(await response.text());
            setHighlightInvalid(true);
        }
    }

    // Handler for error modal overwrite button when showing duplicate error
    async function confirmOverwriteDuplicate() {
        // Convert friendly name into config filename
        const target_filename = friendlyNameToFilename(config.metadata.id);
        // Close error modal, delete existing file, resubmit
        setErrorModalContent({ ...errorModalContent, ["visible"]: false });
        await send_post_request("delete_config", `${target_filename}.json`);
        await submitButton();
    }

    const PrevPageButton = () => {
        return (
            <Button variant="primary" className="mb-4" onClick={prevPage}>
                Back
            </Button>
        );
    };

    const SubmitButton = () => {
        return (
            <Button variant="primary" className="mb-4" onClick={submitButton}>
                Submit
            </Button>
        );
    };

    const NextPageButton = () => {
        return (
            <Button
                variant="primary"
                className="mb-4"
                onClick={nextPage}
                disabled={page === 3}
            >
                Next
            </Button>
        );
    };

    return (
        <>
            <div className="d-flex flex-column vh-100">
                <h1 className="text-center pt-3 pb-4">{document.title}</h1>
                {(() => {
                    switch(page) {
                        case 1:
                            return <Page1 />;
                        case 2:
                            return <Page2 />;
                        case 3:
                            return <Page3 />;
                    }
                })()}
                <div className="d-flex justify-content-between mx-3 mt-auto">
                    <PrevPageButton />
                    {page === 3 ? <SubmitButton /> : null}
                    <NextPageButton />
                </div>
            </div>
            <ApiTargetRuleModal />
            <UploadModal />
            <ErrorModal />
        </>
    );
};

export default EditConfig;
