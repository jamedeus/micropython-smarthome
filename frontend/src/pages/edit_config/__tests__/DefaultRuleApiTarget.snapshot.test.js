import React from 'react';
import EditConfig from '../EditConfig';
import { ConfigProvider } from 'root/ConfigContext';
import { MetadataContextProvider } from 'root/MetadataContext';
import createMockContext from 'src/testUtils/createMockContext';
import { existingConfigContext, apiTargetOptionsContext } from './mockContext';
import { edit_config_metadata } from 'src/testUtils/mockMetadataContext';

describe('DefaultRuleApiTarget', () => {
    let app, user;

    beforeAll(() => {
        // Create mock state objects
        createMockContext('config', existingConfigContext.config);
        createMockContext('api_target_options', apiTargetOptionsContext);
        createMockContext('instance_metadata', edit_config_metadata);
        createMockContext('edit_existing', existingConfigContext.edit_existing);
        createMockContext('target_node_ip', existingConfigContext.IP);
    });

    beforeEach(async () => {
        // Create user, render EditConfig
        user = userEvent.setup();
        app = render(
            <MetadataContextProvider>
                <ConfigProvider>
                    <EditConfig />
                </ConfigProvider>
            </MetadataContextProvider>
        );
    });

    it('matches snapshot when existing rule targets IR Blaster', async () => {
        // Click button to open ApiTargetRuleModal
        await user.click(app.getByRole('button', { name: 'Set rule' }));

        // Get modal, confirm correct options were rendered
        const modal = app.getByText('API Target Rule').parentElement.parentElement;
        expect(modal).toMatchSnapshot();
    });

    it('matches snapshot when existing rule targets device', async () => {
        // Click button to open ApiTargetRuleModal
        await user.click(app.getByRole('button', { name: 'Set rule' }));

        // Get modal, change target instance to device1, change action to turn_on
        const modal = app.getByText('API Target Rule').parentElement.parentElement;
        await user.selectOptions(within(modal).getAllByRole('combobox')[0], 'device1');
        await user.selectOptions(within(modal).getAllByRole('combobox')[1], 'turn_on');

        // Change Off action target instance to device1, change action to turn_off
        await user.click(app.getByText('Off Action'));
        await user.selectOptions(within(modal).getAllByRole('combobox')[0], 'device1');
        await user.selectOptions(within(modal).getAllByRole('combobox')[1], 'turn_off');

        // Confirm correct options were rendered
        expect(modal).toMatchSnapshot();
    });

    it('matches snapshot when self targetting', async () => {
        // Change target node to self
        await user.selectOptions(app.getByLabelText('Target Node:'), '127.0.0.1');

        // Click button to open ApiTargetRuleModal
        await user.click(app.getByRole('button', { name: 'Set rule' }));

        // Get modal, change target instance to device1, change action to turn_on
        const modal = app.getByText('API Target Rule').parentElement.parentElement;
        await user.selectOptions(within(modal).getAllByRole('combobox')[0], 'device1');
        await user.selectOptions(within(modal).getAllByRole('combobox')[1], 'turn_on');

        // Change Off action target instance to device1, change action to turn_off
        await user.click(app.getByText('Off Action'));
        await user.selectOptions(within(modal).getAllByRole('combobox')[0], 'device1');
        await user.selectOptions(within(modal).getAllByRole('combobox')[1], 'turn_off');

        // Confirm correct options were rendered
        expect(modal).toMatchSnapshot();
    });
});
