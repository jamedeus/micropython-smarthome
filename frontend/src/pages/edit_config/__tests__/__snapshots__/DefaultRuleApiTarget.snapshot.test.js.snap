// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultRuleApiTarget matches snapshot when existing rule targets IR Blaster 1`] = `
<div
  class="modal-content"
>
  <div
    class="justify-content-between pb-0 modal-header"
  >
    <button
      class="btn-close  btn btn-link"
      style="visibility: hidden;"
      type="button"
    />
    <h5
      class="modal-title mx-auto"
    >
      API Target Rule
    </h5>
    <button
      class="btn-close  btn btn-link"
      type="button"
    />
  </div>
  <div
    class="d-flex flex-column mx-auto modal-body"
  >
    <button
      class="mx-auto mb-3 btn btn-secondary btn-sm"
      type="button"
    >
      Help
    </button>
    <div
      class="collapse"
    >
      <p
        class="text-center"
      >
        Just like other devices, ApiTargets can be turned on/off by sensors or manually. Instead of effecting a physical device they fire API commands.
      </p>
      <p
        class="text-center"
      >
        Commands are sent to the target node, which can be changed by closing this popup and selecting an option in the "Target Node" dropdown.
      </p>
      <p
        class="text-center"
      >
        The dropdowns below contain all available options for the current target node. Select a command to fire when this device is turned on, and another for when it is turned off.
      </p>
      <p
        class="text-center"
      >
        <button
          class="mx-auto mb-3 btn btn-secondary btn-sm"
          type="button"
        >
          Examples
        </button>
      </p>
      <div
        class="collapse"
      >
        <ul>
          <li>
            Two nodes with motion sensors can work together to cover a large room. Set Sensor1 to target the lights, then set Sensor2 to activate Sensor1 with the 
            <b>
              trigger_sensor
            </b>
             option.
          </li>
          <li>
            The thermostat can change when a door is open or closed. Set up a door sensor targeting this ApiTarget, then select the thermostat and 
            <b>
              set_rule
            </b>
             command below.
          </li>
          <li>
            Any sensor can turn a TV or Air Conditioner on/off by triggering an ApiTarget targeting an 
            <b>
              Ir Blaster
            </b>
            .
          </li>
        </ul>
        <p
          class="text-center"
        >
          <button
            class="mx-auto mb-3 btn btn-secondary btn-sm"
            type="button"
          >
            Close
          </button>
        </p>
      </div>
    </div>
    <div
      class="mx-auto"
    >
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          target instance
        </option>
        <option
          value="device1"
        >
          Air Conditioner (api-target)
        </option>
        <option
          value="sensor1"
        >
          Thermostat (si7021)
        </option>
        <option
          value="ir_key"
        >
          Ir Blaster
        </option>
        <option
          value="ignore"
        >
          Ignore action
        </option>
      </select>
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          command
        </option>
        <option
          value="tv"
        >
          tv
        </option>
        <option
          value="ac"
        >
          ac
        </option>
      </select>
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          key
        </option>
        <option
          value="start"
        >
          start
        </option>
        <option
          value="stop"
        >
          stop
        </option>
        <option
          value="off"
        >
          off
        </option>
      </select>
    </div>
    <div
      aria-label="Set On/Off command"
      class="mx-auto btn-group"
      role="group"
    >
      <input
        autocomplete="off"
        checked=""
        class="btn-check"
        id="show-on-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-on-rule"
        tabindex="0"
      >
        On Action
      </label>
      <input
        autocomplete="off"
        class="btn-check"
        id="show-off-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-off-rule"
        tabindex="0"
      >
        Off Action
      </label>
    </div>
  </div>
  <div
    class="mx-auto modal-footer"
  >
    <div
      id="rule-buttons"
    >
      <button
        class="m-1 btn btn-success"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`DefaultRuleApiTarget matches snapshot when existing rule targets device 1`] = `
<div
  class="modal-content"
>
  <div
    class="justify-content-between pb-0 modal-header"
  >
    <button
      class="btn-close  btn btn-link"
      style="visibility: hidden;"
      type="button"
    />
    <h5
      class="modal-title mx-auto"
    >
      API Target Rule
    </h5>
    <button
      class="btn-close  btn btn-link"
      type="button"
    />
  </div>
  <div
    class="d-flex flex-column mx-auto modal-body"
  >
    <button
      class="mx-auto mb-3 btn btn-secondary btn-sm"
      type="button"
    >
      Help
    </button>
    <div
      class="collapse"
    >
      <p
        class="text-center"
      >
        Just like other devices, ApiTargets can be turned on/off by sensors or manually. Instead of effecting a physical device they fire API commands.
      </p>
      <p
        class="text-center"
      >
        Commands are sent to the target node, which can be changed by closing this popup and selecting an option in the "Target Node" dropdown.
      </p>
      <p
        class="text-center"
      >
        The dropdowns below contain all available options for the current target node. Select a command to fire when this device is turned on, and another for when it is turned off.
      </p>
      <p
        class="text-center"
      >
        <button
          class="mx-auto mb-3 btn btn-secondary btn-sm"
          type="button"
        >
          Examples
        </button>
      </p>
      <div
        class="collapse"
      >
        <ul>
          <li>
            Two nodes with motion sensors can work together to cover a large room. Set Sensor1 to target the lights, then set Sensor2 to activate Sensor1 with the 
            <b>
              trigger_sensor
            </b>
             option.
          </li>
          <li>
            The thermostat can change when a door is open or closed. Set up a door sensor targeting this ApiTarget, then select the thermostat and 
            <b>
              set_rule
            </b>
             command below.
          </li>
          <li>
            Any sensor can turn a TV or Air Conditioner on/off by triggering an ApiTarget targeting an 
            <b>
              Ir Blaster
            </b>
            .
          </li>
        </ul>
        <p
          class="text-center"
        >
          <button
            class="mx-auto mb-3 btn btn-secondary btn-sm"
            type="button"
          >
            Close
          </button>
        </p>
      </div>
    </div>
    <div
      class="mx-auto"
    >
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          target instance
        </option>
        <option
          value="device1"
        >
          Air Conditioner (api-target)
        </option>
        <option
          value="sensor1"
        >
          Thermostat (si7021)
        </option>
        <option
          value="ir_key"
        >
          Ir Blaster
        </option>
        <option
          value="ignore"
        >
          Ignore action
        </option>
      </select>
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          command
        </option>
        <option
          value="enable"
        >
          enable
        </option>
        <option
          value="disable"
        >
          disable
        </option>
        <option
          value="enable_in"
        >
          enable_in
        </option>
        <option
          value="disable_in"
        >
          disable_in
        </option>
        <option
          value="set_rule"
        >
          set_rule
        </option>
        <option
          value="reset_rule"
        >
          reset_rule
        </option>
        <option
          value="turn_on"
        >
          turn_on
        </option>
        <option
          value="turn_off"
        >
          turn_off
        </option>
      </select>
    </div>
    <div
      aria-label="Set On/Off command"
      class="mx-auto btn-group"
      role="group"
    >
      <input
        autocomplete="off"
        checked=""
        class="btn-check"
        id="show-on-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-on-rule"
        tabindex="0"
      >
        On Action
      </label>
      <input
        autocomplete="off"
        class="btn-check"
        id="show-off-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-off-rule"
        tabindex="0"
      >
        Off Action
      </label>
    </div>
  </div>
  <div
    class="mx-auto modal-footer"
  >
    <div
      id="rule-buttons"
    >
      <button
        class="m-1 btn btn-success"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;

exports[`DefaultRuleApiTarget matches snapshot when self targetting 1`] = `
<div
  class="modal-content"
>
  <div
    class="justify-content-between pb-0 modal-header"
  >
    <button
      class="btn-close  btn btn-link"
      style="visibility: hidden;"
      type="button"
    />
    <h5
      class="modal-title mx-auto"
    >
      API Target Rule
    </h5>
    <button
      class="btn-close  btn btn-link"
      type="button"
    />
  </div>
  <div
    class="d-flex flex-column mx-auto modal-body"
  >
    <button
      class="mx-auto mb-3 btn btn-secondary btn-sm"
      type="button"
    >
      Help
    </button>
    <div
      class="collapse"
    >
      <p
        class="text-center"
      >
        Just like other devices, ApiTargets can be turned on/off by sensors or manually. Instead of effecting a physical device they fire API commands.
      </p>
      <p
        class="text-center"
      >
        Commands are sent to the target node, which can be changed by closing this popup and selecting an option in the "Target Node" dropdown.
      </p>
      <p
        class="text-center"
      >
        The dropdowns below contain all available options for the current target node. Select a command to fire when this device is turned on, and another for when it is turned off.
      </p>
      <p
        class="text-center"
      >
        <button
          class="mx-auto mb-3 btn btn-secondary btn-sm"
          type="button"
        >
          Examples
        </button>
      </p>
      <div
        class="collapse"
      >
        <ul>
          <li>
            Two nodes with motion sensors can work together to cover a large room. Set Sensor1 to target the lights, then set Sensor2 to activate Sensor1 with the 
            <b>
              trigger_sensor
            </b>
             option.
          </li>
          <li>
            The thermostat can change when a door is open or closed. Set up a door sensor targeting this ApiTarget, then select the thermostat and 
            <b>
              set_rule
            </b>
             command below.
          </li>
          <li>
            Any sensor can turn a TV or Air Conditioner on/off by triggering an ApiTarget targeting an 
            <b>
              Ir Blaster
            </b>
            .
          </li>
        </ul>
        <p
          class="text-center"
        >
          <button
            class="mx-auto mb-3 btn btn-secondary btn-sm"
            type="button"
          >
            Close
          </button>
        </p>
      </div>
    </div>
    <div
      class="mx-auto"
    >
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          target instance
        </option>
        <option
          value="device1"
        >
          Heater (http-get)
        </option>
        <option
          value="device2"
        >
          Accent lights (pwm)
        </option>
        <option
          value="device3"
        >
          Computer screen (desktop)
        </option>
        <option
          value="device4"
        >
          Stairway lights (tasmota-relay)
        </option>
        <option
          value="device5"
        >
          Overhead lights (dimmer)
        </option>
        <option
          value="device6"
        >
          Air conditioner (api-target)
        </option>
        <option
          value="device7"
        >
          Fan (relay)
        </option>
        <option
          value="device8"
        >
          Lamp (bulb)
        </option>
        <option
          value="device9"
        >
          Bias lights (wled)
        </option>
        <option
          value="sensor1"
        >
          Door switch (switch)
        </option>
        <option
          value="sensor2"
        >
          Temp sensor (dht22)
        </option>
        <option
          value="sensor3"
        >
          Thermostat (si7021)
        </option>
        <option
          value="sensor4"
        >
          Computer activity (desktop)
        </option>
        <option
          value="sensor5"
        >
          Motion (pir)
        </option>
        <option
          value="sensor6"
        >
          Sunrise (dummy)
        </option>
        <option
          value="ir_key"
        >
          Ir Blaster
        </option>
      </select>
      <select
        class="mb-3 modal-dropdown api-command form-select"
      >
        <option
          value=""
        >
          Select 
          command
        </option>
        <option
          value="enable"
        >
          enable
        </option>
        <option
          value="disable"
        >
          disable
        </option>
        <option
          value="enable_in"
        >
          enable_in
        </option>
        <option
          value="disable_in"
        >
          disable_in
        </option>
        <option
          value="set_rule"
        >
          set_rule
        </option>
        <option
          value="reset_rule"
        >
          reset_rule
        </option>
        <option
          value="turn_on"
        >
          turn_on
        </option>
        <option
          value="turn_off"
        >
          turn_off
        </option>
      </select>
    </div>
    <div
      aria-label="Set On/Off command"
      class="mx-auto btn-group"
      role="group"
    >
      <input
        autocomplete="off"
        checked=""
        class="btn-check"
        id="show-on-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-on-rule"
        tabindex="0"
      >
        On Action
      </label>
      <input
        autocomplete="off"
        class="btn-check"
        id="show-off-rule"
        name="radio"
        type="radio"
        value=""
      />
      <label
        class="btn btn-secondary"
        for="show-off-rule"
        tabindex="0"
      >
        Off Action
      </label>
    </div>
  </div>
  <div
    class="mx-auto modal-footer"
  >
    <div
      id="rule-buttons"
    >
      <button
        class="m-1 btn btn-success"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
